lib/service/storage/virtualIndex.ts

```diff
@@ -0,0 +1,190 @@
+  }
+
+  removeInVirtualIndexMap(notification) {
+    this.virtualIndexMap.delete(notification);
+  }
+
+  getPhysicalIndex(name: string): string {
+    if (this.virtualIndexMap.has(name)) {
+      return this.virtualIndexMap.get(name);
+    }
+    return name;
+  }
+
+  isVirtual(name: string): boolean {
+    return this.virtualIndexMap.has(name);
+  }
+
+  getId(index: string, id: string): string {
+    if (this.isVirtual(index)) {
+      return index + id;
+    }
+    return id;
+  }
+
+  randomString(size = 20) {
+    const buf = Buffer.alloc(size);
+    for (let i = 0; i < size; i++) {
+      buf.writeUInt8((Math.random() * 255) & 255, i);
+    }
+    return buf.toString("base64").slice(0, size);
+  }
+
+  getVirtualId(index: string, id: string): string {
+    if (this.isVirtual(index) && id.startsWith(index)) {
+      return id.substring(index.length, id.length);
+    }
+    return id;
+  }
+
+  public async createVirtualIndex(virtualIndex: string, index: string) {
+    global.kuzzle.emit(VirtualIndex.createVirtualIndexEvent, {
+      physical: index,
+      virtual: virtualIndex,
+    });
+
+    this.virtualIndexMap.set(virtualIndex, index);
+    await global.kuzzle.ask(
+      "core:storage:private:document:create",
+      "kuzzle",
+      "virtual-indexes",
+      { physical: index, virtual: virtualIndex },
+      { id: index + virtualIndex }
+    );
+  }
+
+  async removeVirtualIndex(index: string) {
+    const physicalIndex = this.virtualIndexMap.get(index);
+    global.kuzzle.emit(VirtualIndex.deleteVirtualIndexEvent, {
+      virtual: index,
+    });
+    this.virtualIndexMap.delete(index);
+    const id = physicalIndex + index;
+    await global.kuzzle.ask(
+      "core:storage:private:document:delete",
+      "kuzzle",
+      "virtual-indexes",
+      id
+    );
+  }
+
+  async initVirtualIndexList() {
+    //this.virtualIndexMap = new Map<string, string>();
+    let from = 0;
+    let total;
+    //await new Promise(r => setTimeout(r, 2000));
+    //console.log('initVirtualIndexList(sleep)');
+
+    do {
+      const list = await global.kuzzle.ask(
+        "core:storage:private:document:search",
+        "kuzzle",
+        "virtual-indexes",
+        { from, size: 100 }
+      );
+      total = 0;
+      if (list) {
+        total = list.total;
+        for (const hit of list.hits) {
+          this.virtualIndexMap.set(hit._source.virtual, hit._source.physical);
+        }
+        from += 100;
+      }
+    } while (from < total);
+  }
+
+  async buildCollection() {
+    try {
+      await global.kuzzle.ask(
+        "core:storage:private:collection:create",
+        "kuzzle",
+        "virtual-indexes",
+        {
+          mappings: {
+            _meta: undefined,
+            dynamic: "strict",
+            properties: {
+              physical: { type: "keyword" },
+              virtual: { type: "keyword" },
+            },
+          },
+        }
+      );
+    } catch (e) {
+      if (e.status !== 412) {
```

Comment:
You should use the correct error code instead of the error status. They are more specific and also more explicit